### PR TITLE
Product downloadable files

### PR DESCRIPTION
**Pull request by hichamboushaba**
### Description
This PR includes the full work done so far on the product downloadable files.
For a detailed description of the changes, and the testing steps, check the initial PRs:

- PR #2727 (fixes issue #2708)
- PR #2730 (fixes issue #2709)
- PR #2747 (fixes issue #2745)
- PR #2742 (fixes issue #2710)
- PR #2760 (fixes issue #2711)

### Screenshots
**Display Files**
Screen Shot 2020-08-13 at 19 39 11 Screen Shot 2020-08-13 at 19 39 20

**Edit files**
Screen Shot 2020-08-18 at 08 32 17 Screen Shot 2020-08-18 at 08 32 00
update-downloads

**Delete files**
Screen Shot 2020-08-18 at 19 14 07 Screen Shot 2020-08-18 at 19 14 13
delete-files

**Mark file as downloadable**
Screen Shot 2020-08-19 at 15 28 14 Screen Shot 2020-08-19 at 15 28 19

Update release notes:

- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by anitaa1990** [comment]
This PR should not be merged till this [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1663) is merged and the hash updated into this PR but it's ready for review 👍

**Review by 0nko — DISMISSED** [review]
This looks great, good job! 👍 

I just have quick question - since we're merging this to `develop`, should the Add file be hidden until it's functional?

**Comment by hichamboushaba** [comment]
> This looks great, good job! 👍 
> 
> I just have quick question - since we're merging this to `develop`, should the Add file be hidden until it's functional?

Thanks @0nko, regarding your question, the downloadable files item is hidden behind the M4 feature flag, so users won't see the Add file button, and also our goal is to implement it before merging this PR, I'm working on a last PR to implement #2711.

**Comment by ThomazFB** [comment]
hey @anitaa1990! I think this PR is finally ready for a new round, I've tested a lot of scenarios and submitted some fixes here and there. Do you think you can take a look when you have some time available? Let me know what you think and thank you!

**Review by Garance91540 — DISMISSED** [review]
Looking good! 

Is it possible to check the dialog padding so it matches the one on Figma? https://d.pr/i/8ES75C

Could be just my eyes but it looks slightly different

**Review by anitaa1990 — COMMENTED** [review]
Hey @ThomazFB! Thanks so much for working on this! I am not done with my testing yet (lots of changes 😅 ) but I wanted to leave a few comments after testing so that you can get started:

- Keep editing option in Discard dialog is no longer visible.
LEFT: from this PR.  RIGHT: from `develop` branch:
. 

- Update menu button is still visible even after discarding the changes in the Downloadable Files Settings screen.


- Clicking on add image icon in the Product detail screen led to the following crash:

```
2020-10-15 15:44:57.252 7100-7100/com.woocommerce.android E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.woocommerce.android, PID: 7100
    java.lang.IllegalArgumentException: No destination with ID 2131362593 is on the NavController's back stack
        at androidx.navigation.NavController.getBackStackEntry(NavController.java:1293)
        at com.woocommerce.android.ui.products.ProductImagesModule$Companion.provideSavedStateRegistryOwner(ProductImagesModule.kt:28)
        at com.woocommerce.android.ui.products.ProductImagesModule.provideSavedStateRegistryOwner(Unknown Source:2)
        at com.woocommerce.android.ui.products.ProductImagesModule_ProvideSavedStateRegistryOwnerFactory.provideSavedStateRegistryOwner(ProductImagesModule_ProvideSavedStateRegistryOwnerFactory.java:37)
        at com.woocommerce.android.ui.products.ProductImagesModule_ProvideSavedStateRegistryOwnerFactory.get(ProductImagesModule_ProvideSavedStateRegistryOwnerFactory.java:27)
        at com.woocommerce.android.ui.products.ProductImagesModule_ProvideSavedStateRegistryOwnerFactory.get(ProductImagesModule_ProvideSavedStateRegistryOwnerFactory.java:9)
        at com.woocommerce.android.viewmodel.ViewModelFactory_Factory.get(ViewModelFactory_Factory.java:37)
        at com.woocommerce.android.viewmodel.ViewModelFactory_Factory.get(ViewModelFactory_Factory.java:12)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.woocommerce.android.ui.products.ProductImagesFragment$viewModel$2.invoke(ProductImagesFragment.kt:53)
        at com.woocommerce.android.ui.products.ProductImagesFragment$viewModel$2.invoke(ProductImagesFragment.kt:45)
        at com.woocommerce.android.ui.products.ProductImagesFragment$$special$$inlined$navGraphViewModels$3.invoke(NavGraphViewModelLazy.kt:62)
        at com.woocommerce.android.ui.products.ProductImagesFragment$$special$$inlined$navGraphViewModels$3.invoke(Unknown Source:0)
        at androidx.lifecycle.ViewModelLazy.getValue(ViewModelProvider.kt:52)
        at androidx.lifecycle.ViewModelLazy.getValue(ViewModelProvider.kt:41)
        at com.woocommerce.android.ui.products.ProductImagesFragment.getViewModel(Unknown Source:2)
        at com.woocommerce.android.ui.products.ProductImagesFragment.onViewCreated(ProductImagesFragment.kt:85)
        at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:332)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1199)
        at androidx.fragment.app.FragmentManager.addAddedFragments(FragmentManager.java:2236)
        at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2009)
        at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:1965)
        at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:1861)
        at androidx.fragment.app.FragmentManager$4.run(FragmentManager.java:413)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7319)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:934)
```

**Comment by anitaa1990** [comment]
@nbradbury would you also be able to take a look at this PR when you are free? There are a bunch of changes to Product Images (we are reusing the same implementation for uploading downloadable files) and you might have more a bit more context on the product images flow than I do.

**Comment by nbradbury** [comment]
I think this needs to be "access" rather than "accessed." Also, "URL" is lowercase the first time it's mentioned, then uppercase the second. And I'm not sure all users are going to understand what it means to require the URL to be encoded.

Screen Shot 2020-10-16 at 7 07 12 AM

**Comment by nbradbury** [comment]
If I re-order the downloads, the toolbar has an "Update" button. That should be a "Done" button since the changes aren't applied until the user returns to the primary product details screen and taps "Update."

update

**Comment by nbradbury** [comment]
If I go to product detail for a downloadable product and then view the downloads, when I return to product detail the product image no longer shows.

ezgif-5-01d5feafc6e4

**Comment by nbradbury** [comment]
~~The Downloadable Files fragment needs to use the same slide-in/slide-out transition we added in #2977.~~

Ignore that, turns out I had a stale branch :) I see that this has been added.

**Comment by nbradbury** [comment]
The downloadable file list should show a ripple when a file is tapped.

**Comment by nbradbury** [comment]
I'm not sure we should use "Settings" as the menu caption. Maybe "Download settings" instead, to make it clear the settings are for those downloads and not the product or app?

settings

**Comment by nbradbury** [comment]
I'm also not sure that "Delete" should be in the options menu. Perhaps a trash icon in the toolbar instead?

delete

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/downloads/ProductDownloadDetailsViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    private fun updateState(updatedState: ProductDownloadDetailsViewState) {
+        val hasChanges = !navArgs.isEditing || updatedState.fileDraft != navArgs.productFile
+        productDownloadDetailsViewState = updatedState.copy(showDoneButton = hasChanges).validateInput()
+    }
+
+    private fun ProductDownloadDetailsViewState.validateInput(): ProductDownloadDetailsViewState {
+        // The backend only accepts the protocols http and https, so if the url doesn't start with http, we add it
+        // this way, if the url had a different protocl, the validation will fail
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/downloads/ProductDownloadDetailsViewModel.kt]: minor nit: typo in "protocl"

**Comment by nbradbury** [comment]
I'm wondering if this screen should make sure the entered URL is valid? I could imagine a user making a typo while entering it.

bogus

**Comment by nbradbury** [comment]
When I upload a file from the device, it shows up with an empty name. Should we instead set the name to the filename without the path, as the web app does?

name

**Comment by nbradbury** [comment]
I'm not sure about using the word "Delete" to remove a file from a downloadable product. That makes it sound like the file will be deleted from their WP media library, which isn't the case. The wording in this dialog certainly makes it sound like the actual files will be deleted:

delete2

**Comment by nbradbury** [comment]
After tapping "Add File" and choosing a file, the file is uploaded right away using a blocking progress dialog. This is problematic for a few reasons:

- This is inconsistent with how the image upload works, because that uses an ongoing notification to let the user know the progress of the upload

- There's no upload progress shown so the user has no idea how long it will take

- The blocking dialog prevents the user from making any other changes to the product, which can be done when images are being uploaded. For large uploads this is even more of an issue.

upload

**Comment by nbradbury** [comment]
I've finished my first pass at this. No idea this was such a large task!

I left several UI-related comments that @Garance91540 might want to review.

**Comment by Garance91540** [comment]
Hi @nbradbury , thank you for your thorough review!

> I think this needs to be "access" rather than "accessed." Also, "URL" is lowercase the first time it's mentioned, then uppercase the second. And I'm not sure all users are going to understand what it means to require the URL to be encoded.

Good eye! I agree with updating "accessed" with "access" and make URL uppercase on all instances. I'm not sure either what "URLs entered here should already be encoded" mean. Does this mean that they should be valid url? @nbradbury do you have any alternative suggestion?

> If I re-order the downloads, the toolbar has an "Update" button. That should be a "Done" button since the changes aren't applied until the user returns to the primary product details screen and taps "Update."

Sounds good, let's update "Update" with "Done"

> I'm not sure we should use "Settings" as the menu caption. Maybe "Download settings" instead, to make it clear the settings are for those downloads and not the product or app?

The designs actually mention "Download settings", so I agree this needs updating!

> I'm also not sure that "Delete" should be in the options menu. Perhaps a trash icon in the toolbar instead?

This is consistent with the "Delete product" action on the Product Details page (accessible via the More Actions menu). So I think we can keep it as such!

> I'm wondering if this screen should make sure the entered URL is valid? I could imagine a user making a typo while entering it.

I had assumed it was already the case, so yes! I think  for all URL field we should make sure the URL is valid :)

> When I upload a file from the device, it shows up with an empty name. Should we instead set the name to the filename without the path, as the web app does?

In the designs for files without names, I actually displayed "File 1", "File 2".. etc  but showing the filename without the path works better as it's aligned with what the web does.

>I'm not sure about using the word "Delete" to remove a file from a downloadable product. That makes it sound like the file will be deleted from their WP media library, which isn't the case. The wording in this dialog certainly makes it sound like the actual files will be deleted:

We can update this dialog's copy here" https://d.pr/i/6AiFvK

**Comment by Garance91540** [comment]
> After tapping "Add File" and choosing a file, the file is uploaded right away using a blocking progress dialog. This is problematic for a few reasons:
> 
> * This is inconsistent with how the image upload works, because that uses an ongoing notification to let the user know the progress of the upload
> * There's no upload progress shown so the user has no idea how long it will take
> * The blocking dialog prevents the user from making any other changes to the product, which can be done when images are being uploaded. For large uploads this is even more of an issue.
> 
> upload

Thanks for spotting this @nbradbury . The logic and behavior here should be exactly the same as the image uploader screen :😃

**Comment by hichamboushaba** [comment]
Sorry to jump on this, I received the notifications about this discussion on my email, and wanted to give my input 🙂 

> > I'm wondering if this screen should make sure the entered URL is valid? I could imagine a user making a typo while entering it.
>
> I had assumed it was already the case, so yes! I think for all URL field we should make sure the URL is valid :)

The screen checks if the URL entered is valid syntactically (using `PatternsCompat.WEB_URL` regex), to align with what the backend expects and does.
@nbradbury Is your suggestion to extend this to check if the URL actually returns a valid response? just FI the web app doesn't do this.

> > When I upload a file from the device, it shows up with an empty name. Should we instead set the name to the filename without the path, as the web app does?
>
> In the designs for files without names, I actually displayed "File 1", "File 2".. etc but showing the filename without the path works better as it's aligned with what the web does.

After updating the product on the main screen, the backend response will return the filenames, and they will be updated. This state on the screenshot happens only after adding a file, and going back.
We can repeat the logic that the backend does for the file names, but this will be different from what the web app does, as what it does now, is the same as our logic, the file can be entered without a file name, and when clicking on Update, the file name will be populated by the backend response.



> > After tapping "Add File" and choosing a file, the file is uploaded right away using a blocking progress dialog. This is problematic for a few reasons:
> > 
> > * This is inconsistent with how the image upload works, because that uses an ongoing notification to let the user know the progress of the upload
> > * There's no upload progress shown so the user has no idea how long it will take
> > * The blocking dialog prevents the user from making any other changes to the product, which can be done when images are being uploaded. For large uploads this is even more of an issue.
> > 
> > upload
> 
> Thanks for spotting this @nbradbury . The logic and behavior here should be exactly the same as the image uploader screen :😃

This is not as simple as this 🙂, as we need the URL for the next screen, and AFAIK we don't get until the upload is completed, so if we wanted to do like the image uploader, the flow has to be changed.
Or we can, to reduce the effort, just make the progress dialog cancellable, and add a progress bar to it to give the user an idea how long it should take.

**Comment by nbradbury** [comment]
> @nbradbury Is your suggestion to extend this to check if the URL actually returns a valid response? just FI the web app doesn't do this.

Yes, since typos on mobile devices are so easy to make. But this wouldn't have to happen in this PR - just a nice future addition.

> Or we can, to reduce the effort, just make the progress dialog cancellable, and add a progress bar to it to give the user an idea how long it should take.

For the short term I think this would be fine. It's much more user-friendly than a blocking dialog that doesn't show the upload progress.

PS: Excellent work on this feature, it wasn't an easy task!

**Comment by ThomazFB** [comment]
Hey everyone, thank you all for the awesome inputs around this discussion, I'll wrap up all context given here and update the PR ASAP. I've already started with the fixes pointed by Anitaa, after I'm finished with that I'll update all UI related requests.

**Comment by ThomazFB** [comment]
Hello @anitaa1990 and @nbradbury! I've submitted some changes around your reviews, I'll do a summary of what's done and you folks can let me know if something is missing, thank you so much for your time!

Anitaa's review:
1 - The `Keep Editing` button which was missing is now restored
2 - I've fixed the `Update` menu button showing up incorrectly at the Product Details View
3 - ⚠️  The add image icon button is something I've not found yet what's the source of the problem, working on it right now

Nick's review:
1 -  The File URL field description is updated and the typos are fixed, but from the discussion with @Garance91540 I understand that the encoded part is something that we don't have a proper suggestion right now about how to change it, correct?
2 - I've changed the `Update` button to `Done` at the Product Downloads view
3 - The ripple effect issue is fixed at the Product Downloads view
4 - The `Settings` text at the overflow button from the Products Downloads view is now updated to `Download Settings` to match the UI guideline
5 - ⚠️  The trash icon suggestion at the Product Download Details view wasn't submitted since @Garance91540 suggested that we should keep as it is to keep consistency with the whole product feature
6 - Comment typo fixed at the `ProductDownloadDetailViewModel` class
7 - ⚠️  The current URL validation, from what I understood from the previous discussion,  is acceptable and we can improve it as a parallel issue later, correct? Following this, I've disposed of what I started around the validation
8 - ⚠️  The empty filename is an expected behavior following the discussion around this PR since the name will be updated once the file is uploaded to the site backend
9 - The `Deleted` word when disabling the downloadable products toggle was replaced by `Removed`, and following @Garance91540 validation the UI for the dialog was updated and can be checked right here:
screenshot-1603973827416

10 - The download progress didn't get any changes since we agreed on those limitations for the short time


Let me know if I didn't cover any item I should've. and again thanks for the review effort, everyone!

**Comment by ThomazFB** [comment]
@anitaa1990 the crash on the image icon click is now fixed, so I think we pretty much covered I pending items for this PR, so I think it's ready for another round!

Thanks a lot to @nbradbury who gave me the hint about what was going on.

**Review by anitaa1990 — APPROVED** [review]
@ThomazFB the code changes look good and I've tested all the scenarios mentioned in the PR description 🥳 I'm good to merge this. Will wait till @nbradbury approves before merging 👍

**Comment by anitaa1990** [comment]
Oops.. there are also some CircleCI tests that are failing

**Comment by ThomazFB** [comment]
Hey @anitaa1990, I had to just do a rerun at the CircleCI tests because it was caused by an environment failure during the compiler stage, looks like it's just fine now and tests are running smoothly. Thank you so much for reviewing it! 🙇 

About the review, @nbradbury do think you can take a look at the changes requested to see if everything is as it should? Thank you for your time!

**Comment by nbradbury** [comment]
> About the review, @nbradbury do think you can take a look at the changes requested to see if everything is as it should? Thank you for your time!

Sorry - missed the first ping! I'm looking at it now.

**Review by nbradbury — APPROVED** [review]
Looks great! :shipit:

**Pull request merged**